File: A2Z_striver/1 Basics/1.6 String/reverseString/reverseString.py
class Solution:
    # ~ Approach: Two Pointers (Optimal)
    # Analysis: Time O(N) | Space O(1)
    def reverseString(self, s):
        """
        Do not return anything, modify s in-place instead.
        """
        left = 0
        right = len(s) - 1
        
        # - Step: Iterate until pointers meet
        while left < right:
            # - Logic: Swap elements
            # Python allows tuple unpacking for easy swapping
            s[left], s[right] = s[right], s[left]
            
            # - Step: Move pointers inward
            left += 1
            right -= 1

# --- Driver Code ---
if __name__ == "__main__":
    sol = Solution()
    
    # Test Case
    s = ["h","e","l","l","o"]
    sol.reverseString(s)
    
    print(f"Reversed: {s}") 
    # Output: ['o', 'l', 'l', 'e', 'h']

File: A2Z_striver/1 Basics/1.6 String/reverseString/reverseString.cpp
#include <iostream>
#include <vector>
#include <algorithm> // for swap

using namespace std;

// Header Comment Block
// Problem: Reverse String.
// Constraint: In-place O(1) memory.

class Solution {
public:
    // ~ Approach: Two Pointers
    // Analysis: Time O(N) | Space O(1)
    void reverseString(vector<char>& s) {
        int left = 0;
        int right = s.size() - 1;

        // - Step: Iterate until pointers meet
        while (left < right) {
            
            // - Logic: Swap values at pointers
            // Manual swap: char temp = s[left]; s[left] = s[right]; s[right] = temp;
            swap(s[left], s[right]);

            // - Step: Move pointers
            left++;
            right--;
        }
    }
};

int main() {
    Solution sol;
    
    vector<char> s = {'h', 'e', 'l', 'l', 'o'};
    
    sol.reverseString(s);
    
    cout << "Reversed: ";
    for(char c : s) cout << c << " ";
    cout << endl;

    return 0;
}